cpp/src/arrow/type_fwd.h

```diff
@@ -724,4 +724,23 @@ ARROW_EXPORT MemoryPool* default_memory_pool();
 
 constexpr int64_t kDefaultBufferAlignment = 64;
 
+/// \brief EXPERIMENTAL: Device type enum which matches up with C Data Device types
+enum class DeviceAllocationType : char {
+  kCPU = 1,
+  kCUDA = 2,
+  kCUDA_HOST = 3,
+  kOPENCL = 4,
+  kVULKAN = 7,
+  kMETAL = 8,
+  kVPI = 9,
+  kROCM = 10,
+  kROCM_HOST = 11,
+  kEXT_DEV = 12,
+  kCUDA_MANAGED = 13,
+  kONEAPI = 14,
+  kWEBGPU = 15,
+  kHEXAGON = 16,
+};
+constexpr int kDeviceAllocationTypeMax = 16;
+
```

Comment:
kernel.h etc wouldn't need to change if we forward declare here
```suggestion
class DeviceAllocationTypeSet;
```

Reply:
Not really because I'm passing `DeviceAllocationTypeSet` by value. The whole instance fits in a 64-bit word.

Reply:
But I added a forward declaration to it nevertheless.